File: game-server/src/main/java/org/forfun/mmorpg/game/asyncdb/AsyncDbService.java
package org.forfun.mmorpg.game.asyncdb;

import jakarta.annotation.PostConstruct;
import jakarta.annotation.PreDestroy;
import jforgame.commons.persist.DbService;
import jforgame.commons.persist.Entity;
import jforgame.commons.persist.PersistContainer;
import jforgame.commons.persist.QueueContainer;
import jforgame.commons.persist.SavingStrategy;
import org.forfun.mmorpg.game.ServerType;
import org.forfun.mmorpg.game.base.GameContext;
import org.forfun.mmorpg.game.database.user.entity.PlayerEnt;
import org.springframework.stereotype.Service;

/**
 * 异步持久化服务
 */
@Service
public class AsyncDbService implements DbService {


    private PersistContainer playerWorker;

    private PersistContainer commonWorker;

    @PostConstruct
    private void init() {
        SavingStrategy savingStrategy = new EntitySavingStrategy();
        playerWorker = new QueueContainer("playerDbContainer", savingStrategy);
        commonWorker = new QueueContainer("commonDbContainer", savingStrategy);
    }


    @Override
    public void saveToDb(Entity<?> entity) {
        if (GameContext.serverType != ServerType.GAME) {
            // only game server can saving data
            return;
        }
        if (entity instanceof PlayerEnt) {
            playerWorker.receive(entity);
        } else {
            commonWorker.receive(entity);
        }
    }

    @Override
    public void deleteFromDb(Entity<?> entity) {

    }

    @PreDestroy
    public void shutDown() {
        playerWorker.shutdownGraceful();
        commonWorker.shutdownGraceful();
    }

}

File: game-server/src/main/java/org/forfun/mmorpg/game/asyncdb/AsyncDbService.kt
package org.forfun.mmorpg.game.asyncdb

import jakarta.annotation.PostConstruct
import jakarta.annotation.PreDestroy
import jforgame.commons.persist.*
import org.forfun.mmorpg.game.ServerType
import org.forfun.mmorpg.game.base.GameContext
import org.forfun.mmorpg.game.database.user.entity.PlayerEnt
import org.springframework.stereotype.Service

/**
 * 异步持久化服务
 */
@Service
class AsyncDbService : DbService {
    private var playerWorker: PersistContainer? = null

    private var commonWorker: PersistContainer? = null

    @PostConstruct
    private fun init() {
        val savingStrategy: SavingStrategy = EntitySavingStrategy()
        playerWorker = QueueContainer("playerDbContainer", savingStrategy)
        commonWorker = QueueContainer("commonDbContainer", savingStrategy)
    }


    override fun saveToDb(entity: Entity<*>?) {
        if (GameContext.serverType != ServerType.GAME) {
            // only game server can saving data
            return
        }
        if (entity is PlayerEnt) {
            playerWorker!!.receive(entity)
        } else {
            commonWorker!!.receive(entity)
        }
    }

    override fun deleteFromDb(entity: Entity<*>?) {
    }

    @PreDestroy
    override fun shutDown() {
        playerWorker!!.shutdownGraceful()
        commonWorker!!.shutdownGraceful()
    }
}